server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/dlp/DlpDomainRules.java

```diff
@@ -0,0 +1,310 @@
+                return Objects.equals(id, other.id) &&
+                    Objects.equals(matcher, other.matcher);
+            }
+            return false;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(id, matcher);
+        }
+
+    }
+
+    public static class DlpDomainRulesBuilder {
+
+        private final ImmutableMultimap.Builder<Targets.Type, Rule> rules;
+
+        private DlpDomainRulesBuilder() {
+            rules = ImmutableMultimap.builder();
+        }
+
+        public DlpDomainRulesBuilder recipientRule(DLPConfigurationItem.Id id, Pattern pattern) {
+            return rule(Targets.Type.Recipient, id, pattern);
+        }
+
+        public DlpDomainRulesBuilder senderRule(DLPConfigurationItem.Id id, Pattern pattern) {
+            return rule(Targets.Type.Sender, id, pattern);
+        }
+
+        public DlpDomainRulesBuilder contentRule(DLPConfigurationItem.Id id, Pattern pattern) {
+            return rule(Targets.Type.Content, id, pattern);
+        }
+
+        public DlpDomainRulesBuilder rule(Targets.Type type, DLPConfigurationItem.Id id, Pattern regexp) {
+            rules.put(type, toRule(type, id, regexp));
+            return this;
+        }
+
+        private Rule toRule(Targets.Type type, DLPConfigurationItem.Id id, Pattern pattern) {
+            switch (type) {
+                case Sender:
+                    return new Rule(id, new Rule.SenderMatcher(pattern));
+                case Content:
+                    return new Rule(id, new Rule.ContentMatcher(pattern));
+                case Recipient:
+                    return new Rule(id, new Rule.RecipientsMatcher(pattern));
+                default:
+                    throw new IllegalArgumentException("unexpected value");
+            }
+        }
+
+        public DlpDomainRules build() {
+            ImmutableMultimap<Targets.Type, Rule> rules = this.rules.build();
+            Preconditions.checkState(!containsDuplicateIds(rules), "Rules should not contain duplicated `id`");
+            return new DlpDomainRules(rules.values());
+        }
+
+        private boolean containsDuplicateIds(ImmutableMultimap<Targets.Type, Rule> rules) {
+            return
+                Stream.of(Targets.Type.values())
+                    .map(rules::get)
+                    .anyMatch(this::containsDuplicateIds);
+        }
+
+        private boolean containsDuplicateIds(ImmutableCollection<Rule> rules) {
+            long distinctIdCount = rules.stream()
+                .map(Rule::id)
+                .distinct()
+                .count();
+            return distinctIdCount != rules.size();
+        }
+
+    }
+
+    private final ImmutableCollection<Rule> rules;
+
+    private DlpDomainRules(ImmutableCollection<Rule> rules) {
+        this.rules = rules;
+    }
+
+    public Optional<DLPConfigurationItem.Id> match(Mail mail) {
+        return rules.stream()
+            .flatMap(rule -> applyRule(rule, mail))
```

Comment:
can be shortened by `.filter(rule -> rule.match(mail))`

Reply:
good catch, thanks

Reply:
fixed